### PR TITLE
Support module aliasing through assignment

### LINKED ISSUE #610: Module aliases
When using numpy through an alias, pythran raises an exception.
Here is an example code:

``` python
import numpy as np
np2=np
#pythran export ops_numpy_unary(uint32[],uint32[],uint32[])
def ops_numpy_unary(a, b, c): 
    return np2.sqrt(a+b*c)
```

The exception is (backtrace extract):

```
  File "/home/aguinet/.virtualenvs/pythran/lib/python2.7/site-packages/pythran-0.7.4.post1-py2.7.egg/pythran/transformations/normalize_method_calls.py", line 189, in visit_Call                                                                                                                                             
    node.func.value, _ = rec(node.func.value, MODULES)                                                                                                                                                                                                                                                                       
  File "/home/aguinet/.virtualenvs/pythran/lib/python2.7/site-packages/pythran-0.7.4.post1-py2.7.egg/pythran/transformations/normalize_method_calls.py", line 178, in rec                                                                                                                                                    
    assert isinstance(path, (ast.Name, ast.Attribute)), err                                                                                                                                                                                                                                                                  
AssertionError: Function path is chained attributes and name
```

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Fixes #610

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
@pbrunet , @aguinet this PR still has a few validation issue, but the ideas are there, you can comment / review !

**Comment by serge-sans-paille** [comment]
travis is ok, ready4review!

**Comment by serge-sans-paille** [comment]
@pbrunet : please reaaaad me!

**Comment by pbrunet** [comment]
Sorry, I did the review but I didn't notice github ask for an "approval" before submitting the review. It was pending for weeks...

File: pythran/transformations/normalize_method_calls.py [hunk 1]
```diff
@@ -49,27 +52,42 @@ def visit_Module(self, node):
...
-        for t in node.targets:
-            if isinstance(t, ast.Name):
-                self.imports.discard(t.id)
-        return n
+        # aliasing between modules
+        if isinstance(node.value, ast.Name) and node.value.id in self.imports:
+            for t in node.targets:
+                if isinstance(t, ast.Name):
+                    self.imports[t.id] = self.imports[node.value.id]
+            return None
```
Comment (serge-sans-paille) [inline pythran/transformations/normalize_method_calls.py]: mmmh, I don't know, I'll check and make sure this (strange) test case works

File: pythran/transformations/remove_nested_functions.py [hunk 2]
```diff
@@ -24,7 +25,14 @@ def visit_FunctionDef(self, node):
...
         former_name = node.name
-        new_name = "pythran_{0}".format(former_name)
+        seed = 0
+        new_name = "pythran_{}{}"
+
+        while new_name.format(former_name, seed) in self.identifiers:
+            seed += 1
+
+        new_name = new_name.format(former_name, seed)
+        self.identifiers.add(new_name)
```
Comment (serge-sans-paille) [inline pythran/transformations/remove_nested_functions.py]: will do, in a separate commit (but in this PR)

File: pythran/types/types.py [hunk 3]
```diff
@@ -403,13 +402,13 @@ def visit_Call(self, node):
                 bounded_name = a0.id
                 # by construction of the bind construct
                 assert len(self.strict_aliases[a0]) == 1
-                bounded_function = next(iter((self.strict_aliases[a0])))
+                bounded_function = list(self.strict_aliases[a0])[0]
```
Comment (serge-sans-paille) [inline pythran/types/types.py]: yes, why?

File: pythran/pythonic/__builtin__/pythran/or_.hpp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  namespace __builtin__
+  {
+
+    namespace pythran
+    {
+
+      template <class T0, class T1>
+      types::lazy_combined_t<T0, T1> or_(T0 &&v0, T1 &&v1)
+      {
+        auto &&val0 = std::forward<T0>(v0)();
```
Comment (serge-sans-paille) [inline pythran/pythonic/__builtin__/pythran/or_.hpp]: yes. But genericity makes sure you don't need to worry about the callsite :-)

File: pythran/backend.py [hunk 5]
```diff
@@ -1137,8 +1137,11 @@ def visit_IfExp(self, node):
         test = self.visit(node.test)
         body = self.visit(node.body)
         orelse = self.visit(node.orelse)
-        return ("(pythonic::__builtin__::functor::bool_{{}}"
-                "({0}) ? {1} : {2})".format(test, body, orelse))
+
+        return ("pythonic::__builtin__::pythran::ifexp((bool){0}, {1}, {2})"
```
Comment (serge-sans-paille) [inline pythran/backend.py]: why? gmp expression don't overload the operator bool() ?

Comment (pbrunet) [inline pythran/backend.py]: Looking at the implementation of ``__builtin__::bool_``, I would say it doesn't but may be I am wrong.

Comment (serge-sans-paille) [inline pythran/backend.py]: Tested: you're wrong :-)

**Comment by serge-sans-paille** [comment]
@pbrunet bugs and enhancements (except the ``get unique id stuff`` added), less bugs, thanks 4 the review!

**Comment by serge-sans-paille** [comment]
@pbrunet I rebased against your last commits and pushed the revision

File: pythran/tests/test_normalize_methods.py [hunk 6]
```diff
@@ -11,11 +11,71 @@ class TestBase(TestEnv):
...
+                        import math as ma2
+                        m = ma
+                        def mab():
+                            return m.cos(c) + ma2.cos(c)
+                        return mab()""",
+                      1.,
+                      module_alias4=[float])
+
+    def test_module_alias5(self):
+        self.run_test("import math as m2\ndef module_alias5(math): m = m2 ; return m.cos(math)", 1., module_alias5=[float])
```
Comment (pbrunet) [inline pythran/tests/test_normalize_methods.py]: did you do it?

File: pythran/optimizations/constant_folding.py [hunk 7]
```diff
@@ -31,7 +31,10 @@ def __init__(self):
 
     def prepare(self, node, ctx):
         assert isinstance(node, ast.Module)
-        self.env = {'__builtin__': __import__('__builtin__')}
+        self.env = {
+            '__builtin__': __import__('__builtin__'),
+            mangle('__builtin__'): __import__('__builtin__'),
```
Comment (serge-sans-paille) [inline pythran/optimizations/constant_folding.py]: it does not, you're correct!

File: pythran/tests/test_normalize_methods.py (hunk 6 above)
Comment (serge-sans-paille) [inline pythran/tests/test_normalize_methods.py]: no, that's actually inspired by one of **your** remarks :-)

**Comment by serge-sans-paille** [comment]
@pbrunet thanks for spotting the error in constant_folding ! I'm impressed by the quality of your review ;-)

The code is fixed, let's wait for travis!

File: pythran/tests/test_normalize_methods.py (hunk 6 above)
Comment (pbrunet) [inline pythran/tests/test_normalize_methods.py]: It doesn't mean it is not an obfuscated piece of code :)

**Comment by serge-sans-paille** [comment]
@pbrunet up to you to merge this one :-)

**Pull request merged**